File: src/simulation/culture/aspect/ConverseWrapper.java
package simulation.culture.aspect;

import simulation.culture.aspect.dependency.AspectDependencies;
import simulation.culture.aspect.dependency.LineDependency;
import simulation.culture.group.AspectCenter;
import simulation.space.resource.Resource;
import simulation.space.resource.ResourcePack;
import simulation.space.resource.tag.ResourceTag;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Objects;
import java.util.stream.Collectors;

/**
 * Special Aspect which wraps around another aspect and resource and returns application
 * of this aspect to the resource.
 */
public class ConverseWrapper extends Aspect {
    /**
     * Aspect which is applied to the resource.
     */
    public Aspect aspect;
    public Resource resource;

    public ConverseWrapper(Aspect aspect, Resource resource) {
        super(
                new AspectCore(
                        aspect.getName() + "On" + resource.getBaseName(),
                        getReducedTags(resource, aspect),
                        new ArrayList<>(aspect.getRequirements()),
                        new ArrayList<>(),
                        false
                        ),
                new AspectDependencies(new HashMap<>())
        );
        this.aspect = aspect;
        this.resource = resource.copy();
    }

    @Override
    public void swapDependencies(AspectCenter aspectCenter) {
        super.swapDependencies(aspectCenter);
        aspect = aspectCenter.getAspectPool().getValue(aspect);
    }

    private static List<ResourceTag> getReducedTags(Resource resource, Aspect aspect) {
        List<ResourceTag> result = new ArrayList<>();
        List<ResourceTag> allTags = resource.applyAspect(aspect).stream()
                .flatMap(r -> r.getTags().stream())
                .collect(Collectors.toList());
        for (ResourceTag tag: allTags) {
            if (!result.contains(tag)) {
                result.add(tag);
            } else if (result.get(result.indexOf(tag)).level < tag.level) {
                result.remove(tag);
                result.add(tag);
            }
        }
        return result;
    }

    @Override
    protected ResourcePack getAllProducedResources() {
        return new ResourcePack(resource.applyAspect(aspect));
    }

    public List<Resource> getResult() {
        return resource.applyAspect(aspect);
    }

    @Override
    public AspectResult use(AspectController controller) {
        try {
            aspect.markAsUsed();
            return super.use(controller);
        } catch (Exception e) {
            throw new RuntimeException("");
        }
    }

    @Override
    public boolean isDependenciesOk(AspectDependencies dependencies) {
        return getRequirements().size() + 1 == dependencies.getSize();
    }

    @Override
    public boolean isValid() {
        if (resource.getGenome().isResisting() && aspect.getName().equals("Take")) {
            return false;
        }
        if (!resource.hasApplicationForAspect(aspect)) {
            return false;
        }
        return true;
    }

    @Override
    public boolean canTakeResources() {
        return aspect.getName().equals("Take") ||
                aspect.getName().equals("Killing") ||
                aspect.getName().equals("TakeApart");
    }

    @Override
    public ConverseWrapper copy(AspectDependencies dependencies) {
        ConverseWrapper copy = new ConverseWrapper(
                aspect,
                resource
        );
        copy.initDependencies(dependencies);
        try {
            copy.canInsertMeaning = dependencies.getMap().get(ResourceTag.phony()).stream().anyMatch(dependency ->
                    dependency instanceof LineDependency &&
                            ((LineDependency) dependency).getConverseWrapper().canInsertMeaning);
            return copy;
        } catch (Exception e) {
            return null;
        }
    }

    @Override
    public int hashCode() {
        return Objects.hash(aspectCore.getName(), aspect, resource);
    }
}

File: src/simulation/culture/aspect/ConverseWrapper.kt
package simulation.culture.aspect

import simulation.culture.aspect.dependency.AspectDependencies
import simulation.culture.aspect.dependency.LineDependency
import simulation.culture.group.GroupError
import simulation.culture.group.centers.AspectCenter
import simulation.space.resource.Resource
import simulation.space.resource.tag.ResourceTag
import java.util.*

/**
 * Special Aspect which wraps around another aspect and resource and returns application
 * of this aspect to the resource.
 */
open class ConverseWrapper(var aspect: Aspect, resource: Resource) : Aspect(
        AspectCore(
                aspect.name + "On" + resource.baseName,
                getReducedTags(resource, aspect),
                ArrayList(aspect.requirements),
                ArrayList(),
                false,
                aspect.core.resourceExposed,
                aspect.core.standardComplexity,
                aspect.core.sideComplexities
        ),
        AspectDependencies(mutableMapOf())
) {
    var resource: Resource = resource.copy()

    override fun swapDependencies(aspectCenter: AspectCenter) {
        super.swapDependencies(aspectCenter)
        aspect = aspectCenter.aspectPool.getValue(aspect)
    }

    override val producedResources: List<Resource>
        get() = resource.applyAspect(aspect)

    override fun use(controller: AspectController) = try {
        val result = super.use(controller)
        if (result.isFinished) {
            aspect.markAsUsed()
        }
        result
    } catch (e: Exception) {
        throw GroupError("Error in using Converse Wrapper $name")
    }

    override fun isDependenciesOk(dependencies: AspectDependencies) =
            requirements.size + 1 == dependencies.size

    override val isValid: Boolean
        get() {
            if (resource.genome.isResisting && aspect.name == "Take")
                return false
            return resource.hasApplicationForAspect(aspect)
        }

    override fun canTakeResources() =
            aspect.name == "Take" || aspect.name == "Killing" || aspect.name == "TakeApart"

    override fun copy(dependencies: AspectDependencies): ConverseWrapper {
        val copy = ConverseWrapper(
                aspect,
                resource
        )
        copy.initDependencies(dependencies)
        return try {
            copy.canInsertMeaning = dependencies.map.getValue(ResourceTag.phony()).any {
                it is LineDependency && it.converseWrapper.canInsertMeaning
            }
            copy
        } catch (e: Exception) {
            throw GroupError("Wrong dependencies")
        }
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false
        if (!super.equals(other)) return false

        other as ConverseWrapper

        if (aspect != other.aspect) return false
        if (resource != other.resource) return false
        if (name != other.name) return false

        return true
    }

    override fun hashCode(): Int {
        var result = super.hashCode()
        result = 31 * result + aspect.hashCode()
        result = 31 * result + resource.hashCode()
        result = 31 * result + name.hashCode()
        return result
    }
}

private fun getReducedTags(resource: Resource, aspect: Aspect): List<ResourceTag> {
    val result: MutableList<ResourceTag> = ArrayList()
    val allTags = resource.applyAspect(aspect).flatMap { it.tags }
    for (tag in allTags) {
        if (!result.contains(tag))
            result.add(tag)
        else if (result[result.indexOf(tag)].level < tag.level) {
            result.remove(tag)
            result.add(tag)
        }
    }
    return result
}